Zend/zend_execute.c

```diff
@@ -1177,8 +1177,14 @@ static zend_always_inline bool zend_check_type_slow(
 	}
 
 	const uint32_t type_mask = ZEND_TYPE_FULL_MASK(*type);
-	if ((type_mask & MAY_BE_CALLABLE) &&
-		zend_is_callable(arg, is_internal ? IS_CALLABLE_SUPPRESS_DEPRECATIONS : 0, NULL)) {
+	if (
+		(type_mask & MAY_BE_CALLABLE)
+		&& (
+			/* Fast check for closures. */
+			EXPECTED(Z_OBJCE_P(arg) == zend_ce_closure)
```

Comment:
I don't think `arg` is guaranteed to be an object at this point. Maybe this can be combined with the `Z_TYPE_P(arg) == IS_OBJECT)` above.

Reply:
I don't think you can combine it with the one above as a type declaration that only contains built-in types won't be complex.

Reply:
I was referring to something like:

```c
if (Z_TYPE_P(arg) == IS_OBJECT) {
    if ((type_mask & MAY_BE_CALLABLE) && EXPECTED(Z_OBJCE_P(arg) == zend_ce_closure)) {
        return true;
    } else if (ZEND_TYPE_IS_COMPLEX(*type)) {
        ...
    }
}
```

Not sure if that's better

Reply:
I think it is reasonable to have `if(EXPECTED(Z_TYPE_P(arg) == IS_OBJECT) {}` first, as to get to this point you need to cross `if (EXPECTED(ZEND_TYPE_CONTAINS_CODE(*type, Z_TYPE_P(arg)))) {` from `zend_check_type()`.

Meaning that we are either dealing with an object, a callable, or a scalar type that needs to be coerced which depends on strict_type.

Reply:
Good point. The may also still check for `ZEND_TYPE_IS_COMPLEX(*type)` first to avoid any degradation of complex types.

Reply:
I think the logic should be:
- Check for a name type
- Check for static
- Check for closure
- Check for list type

Reason being is that the complex list type case is already guarded by `UNEXPECTED(ZEND_TYPE_HAS_LIST(*type))` which IIRC is quite a punishing hint to the CPU.

Reply:
Adjusted according to Gina's suggestion.

Reply:
Is static really more common than callable?

Reply:
> Is static really more common than callable?

I'm seeing `static` used quite a bit for return types. The `closure` check is probably cheaper, though. No strong opinion here. Can reorder if desired (ideally after someone else verified the numbers, because I don't really trust the results on my CPU).

Reply:
Same. My benchmarks are usually all over the place... I have yet to find a setting for me that works well.

Reply:
I guess we have static after the quick Closure check, especially as we could now transform `static` to the class name for final classes.